`impl Hash for Robj`

CGMossa opened the pull request:
This is based on previous PR start in #850.

I used codex to implement the rest of this PR. I reviewed the code. I checked other crates, like https://lib.rs/crates/ordered-float, and inspected the used impl. I think this is fairly solid. 

### Misc

I've extracted unrelated things from this PR, into other PRs. Please review these first
- [x] #1005
- [x] #1004
- [x] #1003

JosiahParry commented on extendr-api/src/robj/mod.rs:
```diff
@@ -107,10 +110,241 @@ mod tests;
+fn hash_str_value<H: Hasher>(value: &str, state: &mut H) {
+    use crate::na::CanBeNA;
+    let is_na = value.is_na();
+    is_na.hash(state);
```
what is this?

CGMossa replied:
It is gone 📦

JosiahParry reviewed (changes requested):
@CGMossa would you mind adding comments above each enum variant describing **what** is being hashed?

CGMossa commented:
Out standing task: The API for closures are not available in oldrel, but used by this. Some feature gating is necessary

CGMossa commented on extendr-ffi/build.rs:
```diff
@@ -215,9 +215,9 @@ fn main() -> Result<(), Box<dyn Error>> {
+    // S4 non-vector was introduced in R 4.4
     if (r_paths.version.major, r_paths.version.minor) >= (4, 4) {
-        println!("cargo:rustc-cfg=r_4_4")
+        println!("cargo:rustc-cfg=use_objsxp")
```
#1010

CGMossa committed:
feat: impl Hash and Eq for Robj (#992)

Squashed rebase onto origin/main. Original branch had 28 WIP commits;
collapsed to a single commit to resolve conflicts cleanly against main's
post-fork changes (Environment::global() vs global_env(), explicit
Result types, conditions module, etc.).
